Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1252,9 +1309,54 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
 void WebExtensionContext::performTasksAfterBackgroundContentLoads()
 {
-    // FIXME: <https://webkit.org/b/246483> Implement. Fire setup and install events (if needed), perform pending actions, schedule non-persistent page to unload (if needed), etc.
+    // FIXME: <https://webkit.org/b/246483> Implement. Fire setup and install events (if needed), etc.
+
+    for (auto& event : m_actionsToPerformAfterBackgroundPageLoads)
+        event();
+    m_actionsToPerformAfterBackgroundPageLoads.clear();
 
     saveBackgroundPageListenersToStorage();
+
+    scheduleBackgroundPageToUnload();
+}
+
+void WebExtensionContext::fireEvents(EventListenerTypeSet types, CompletionHandler<void()>&& completionHandler)
+{
+    if (extension().backgroundContentIsPersistent()) {
+        completionHandler();
+        return;
+    }
+
+    bool backgroundPageListensToAtLeastOneEvent = false;
+    for (auto& type : types) {
+        if (m_backgroundPageListeners.contains(type)) {
+            backgroundPageListensToAtLeastOneEvent = true;
+            break;
+        }
+    }
+
+    if (!m_backgroundWebView && backgroundPageListensToAtLeastOneEvent) {
+        queueEventToFireAfterBackgroundPageLoads(WTFMove(completionHandler));
+        loadBackgroundWebView();
+        return;
+    }
+
+    bool backgroundPageIsLoading = m_actionsToPerformAfterBackgroundPageLoads.size();
+    if (backgroundPageIsLoading)
+        queueEventToFireAfterBackgroundPageLoads(WTFMove(completionHandler));
+    else {
+        if (backgroundPageListensToAtLeastOneEvent)
+            scheduleBackgroundPageToUnload();
+
+        completionHandler();
+    }
+}
+
+void WebExtensionContext::queueEventToFireAfterBackgroundPageLoads(CompletionHandler<void()> &&completionHandler)
```

Comment:
I'm trying to avoid the term "BackgroundPage" since it can be a service worker now. This comment applies to all the instances in this PR.

`queueEventToFireAfterBackgroundContentLoads`